File: 13-03-18/python/01-Hamiltonian-and-Lagrangian.py
n = int(input())
ar = [int(x) for x in input().split()]
maxi = {}

maxi[n-1] = ar[n-1]
for i in range(n-2,-1,-1):
    maxi[i] = max(maxi[i+1],ar[i])

for i in range(0,n):
    if(ar[i]>=maxi[i]):
        print(ar[i],end=' ')

File: 13-03-18/java/01-Hamiltonian-and-Lagrangian.java
import java.util.*;

class TestClass {
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int n=input.nextInt();
        int ar[]=new int[n];
        int maxi[]=new int[n];
        
        for(int i=0;i<n;i++)
            ar[i]=input.nextInt();
        
        maxi[n-1]=ar[n-1];
        for(int i=n-2;i>=0;i--)
            maxi[i]=Math.max(maxi[i+1],ar[i]);
        
        for(int i=0;i<n;i++)
            if(ar[i]>=maxi[i])
                System.out.print(ar[i]+" ");
    }
}
